File: nexus/nexus.py
from math import exp, log
from nexus.blocklocation import BlockLocation

class Nexus(BlockLocation):
   
   def __init__(self, minfluence, maxfluence, radius, diffuse):
      super().__init__(0, 0, 0, 'world')
      
      self.minfluence = minfluence
      self.maxfluence = maxfluence
      self.radius = radius
      self.radius2 = radius**2
      self.diffuse = diffuse
      
      self.a = maxfluence - minfluence + diffuse
      self.b = minfluence - diffuse
      self.c = (self.radius2) / log(diffuse / self.a)
   
   def influence(self, loc=None, dist=None, dist2=0):
      '''Returns the influence the Nexus has on a location. A distance, or
      distance squared can also be provided.
      
      This result of this function is based on the Gaussian function.
      '''
      if loc:
         dist2 = self.dist2(loc)
      elif dist:
         dist2 = dist**2
         
      if dist2 > self.radius2:
         return 0
      else:
         return self.a * exp(dist2 / self.c) + self.b

   def upkeep(self):
      pass

File: src/main/java/com/nashorsmind/nexus/Nexus.java
package com.nashorsmind.nexus;

import org.bukkit.Location;

public class Nexus extends BlockLocation {

    private int minfluence, maxfluence;
    private int radius, radiusSquared;
    private double diffuse;
    
    private double a, b, c;
    
    public Nexus(Location location, int min, int max, int radius, double diffuse) {
        super(location);
        this.minfluence = min;
        this.maxfluence = max;
        this.radius = radius;
        this.radiusSquared = radius * radius;
        this.diffuse = diffuse;
        
        this.calculateConstants();
    }
    
    private void calculateConstants() {
        this.a = this.maxfluence - this.minfluence + this.diffuse;
        this.b = this.minfluence - this.diffuse;
        this.c = this.radiusSquared / Math.log(diffuse / this.a);
    }
    
    public double influence(Location location) {
        double distSquared = this.distSquared(location);
        
        if (distSquared > this.radiusSquared) {
            return 0;
        } else {
            return this.a * Math.exp(distSquared / this.c) + this.b;
        }
    }
    
    @Override
    public String toString() {
        return String.format(
            "Nexus=[%d,%d,%d](%d,%d,%d,%f)",
            this.getX(), this.getY(), this.getZ(),
            this.minfluence, this.maxfluence,
            this.radius,
            this.diffuse
        );
    }

}
